File: packages/typescript/src/adapters/claudeCode.ts
import { EnvironmentSpec } from "../spec.js";
import { AgentAdapter, ensureBinAvailable } from "./base.js";
import { readClaudeCredentials } from "./credentials.js";

export interface ClaudeCodeOptions {
  model?: string;
  reasoningEffort?: string;
  loadCredentials?: boolean;
  extraArgv?: string[];
}

export function ClaudeCode(opts: ClaudeCodeOptions = {}): AgentAdapter {
  const loadCredentials = opts.loadCredentials ?? true;
  const adapter: AgentAdapter = {
    id: "claude-code",
    cliBin: "claude",
    configEnvVar: "CLAUDE_CONFIG_DIR",
    async ensureAvailable() {
      await ensureBinAvailable(this.id, this.cliBin);
    },
    async buildSpec(): Promise<EnvironmentSpec> {
      const seedFiles: Record<string, string> = {};
      const credentials: Record<string, string> = {};
      const fileModes: Record<string, number> = {};
      if (loadCredentials) {
        credentials[".credentials.json"] = await readClaudeCredentials();
        fileModes[".credentials.json"] = 0o600;
      } else {
        seedFiles[".claude.json"] = JSON.stringify({ hasCompletedOnboarding: true });
      }
      return new EnvironmentSpec({
        adapterId: "claude-code",
        envOverrides: { CLAUDE_CONFIG_DIR: "$EPHEMERAL_HOME" },
        seedFiles,
        credentials,
        fileModes,
        prefix: "agent-venv-claude-",
      });
    },
    buildArgv(_prompt: string): string[] {
      if (!opts.model) {
        throw new Error("ClaudeCode.buildArgv requires options.model");
      }
      const argv = [
        "claude",
        "--print",
        "--model",
        opts.model,
        "--output-format",
        "stream-json",
        "--verbose",
        "--dangerously-skip-permissions",
      ];
      if (opts.reasoningEffort) argv.push("--effort", opts.reasoningEffort);
      if (opts.extraArgv) argv.push(...opts.extraArgv);
      return argv;
    },
  };
  return adapter;
}

File: packages/rust/src/adapters/claude_code.rs
use super::base::AgentAdapter;
use super::credentials::read_claude_credentials;
use crate::error::Result;
use crate::spec::EnvironmentSpec;

pub struct ClaudeCode {
    pub model: Option<String>,
    pub reasoning_effort: Option<String>,
    pub extra_argv: Vec<String>,
    pub load_credentials: bool,
}

impl ClaudeCode {
    pub fn new() -> Self {
        Self {
            model: None,
            reasoning_effort: None,
            extra_argv: Vec::new(),
            load_credentials: true,
        }
    }
    pub fn model(mut self, m: impl Into<String>) -> Self {
        self.model = Some(m.into());
        self
    }
    pub fn load_credentials(mut self, yes: bool) -> Self {
        self.load_credentials = yes;
        self
    }
    pub fn reasoning_effort(mut self, e: impl Into<String>) -> Self {
        self.reasoning_effort = Some(e.into());
        self
    }
}

impl Default for ClaudeCode {
    fn default() -> Self {
        Self::new()
    }
}

impl AgentAdapter for ClaudeCode {
    fn id(&self) -> &str {
        "claude-code"
    }
    fn cli_bin(&self) -> &str {
        "claude"
    }
    fn config_env_var(&self) -> &str {
        "CLAUDE_CONFIG_DIR"
    }

    fn build_spec(&self) -> Result<EnvironmentSpec> {
        let mut builder = EnvironmentSpec::builder()
            .adapter_id("claude-code")
            .env_override("CLAUDE_CONFIG_DIR", "$EPHEMERAL_HOME")
            .prefix("agent-venv-claude-");
        if self.load_credentials {
            let creds = read_claude_credentials()?;
            builder = builder
                .credential(".credentials.json", creds)
                .file_mode(".credentials.json", 0o600);
        } else {
            builder = builder.seed_file(".claude.json", "{\"hasCompletedOnboarding\":true}");
        }
        Ok(builder.build())
    }

    fn build_argv(&self, _prompt: &str, _workspace: Option<&std::path::Path>) -> Vec<String> {
        let mut argv: Vec<String> = vec![
            "claude".into(),
            "--print".into(),
            "--model".into(),
            self.model.clone().unwrap_or_default(),
            "--output-format".into(),
            "stream-json".into(),
            "--verbose".into(),
            "--dangerously-skip-permissions".into(),
        ];
        if let Some(e) = &self.reasoning_effort {
            argv.push("--effort".into());
            argv.push(e.clone());
        }
        argv.extend(self.extra_argv.iter().cloned());
        argv
    }
}
